EIP-4758: Deactivate selfdestruct (#4758)
ethereum/EIPs: EIPS/eip-4758.md

---
eip: 4758
title: Deactivate SELFDESTRUCT

> description: Deactivate SELFDESTRUCT by changing it to SENDALL, which does recover all funds to the caller but does not delete any contract code or storage.

Reviewer: Description exceeds 140 character max.

author: Guillaume Ballet (@gballet), Vitalik Buterin (@vbuterin), Dankrad Feist (@dankrad)

> discussions-to: https://github.com/ethereum/EIPs/issues/4759

Reviewer 2: Discussions-to links should be to threads at https://ethereum-magicians.org/

status: Draft
type: Standards Track
category: Core
created: 2022-02-03

> requires: 

Reviewer 2: ```suggestion
```

---

> # Deactivating SELFDESTRUCT

Reviewer: ```suggestion
```

## Summary

> This EIP renames the `SELFDESCRUCT` opcode to `SENDALL`, and replaces its functionality. The new functionality will be only to send all Ether in the account to the caller.

Reviewer 3: ```suggestion
This EIP renames the `SELFDESTRUCT` opcode to `SENDALL`, and replaces its functionality. The new functionality will be only to send all Ether in the account to the caller.
```

> # Deactivating SELFDESTRUCT
> 
> ## Summary
> 
> This EIP renames the `SELFDESTRUCT` opcode to `SENDALL`, and replaces its functionality. The new functionality will be only to send all Ether in the account to the caller.
> 

Reviewer 2: ```suggestion
## Abstract

This EIP renames the `SELFDESTRUCT` opcode to `SENDALL`, and replaces its functionality. The new functionality will be only to send all Ether in the account to the caller.

```

## Motivation

> The `SELFDESTRUCT` opcode requires large changes to the state of an account, in particular removing all code and storage. This will not be possible in the future with Verkle trees: Each account will be stored in many different account keys, which will not be obviously connected to the root account (see [Verkle tree EIP](https://notes.ethereum.org/5HDhQXstTaKtVqVbS7S9yw)).

Reviewer 2: ```suggestion
The `SELFDESTRUCT` opcode requires large changes to the state of an account, in particular removing all code and storage. This will not be possible in the future with Verkle trees: Each account will be stored in many different account keys, which will not be obviously connected to the root account.
```
No external links in EIPs, they should be self-contained and contain all necessary information inline.  Note: The motivation doesn't need to a be a fully convincing argument, just give people the gist of "why" and a starting point for them to research more if they want.  If you want to include links, I recommend adding them to the discussions-to link.

This EIP implements this change. Applications that only use SELFDESTRUCT to retrieve funds will still work.

## Specification

>  * The `SELFDESTRUCT` opcode is renamed to `SENDALL`, and now only immediately moves all ETH in the account to the target; it no longer destroys code or storage or alters the nonce

Reviewer 4: Perhaps it also deserves explicit mention here that a `SENDALL` where the target is the invoker no longer burns the ETH in the account.

 * All refunds related to `SELFDESTRUCT` are removed

## Rationale

> Getting rid of the `SELFDESTRUCT` opcode has been considered in the past (for arguments see [here](https://hackmd.io/@vbuterin/selfdestruct)), and there are currently no strong reasons to use it. Disabling it will be a requirement for statelessness.

Reviewer 2: Please include the rationale inline here, not in an external link.  You don't need to provide a complete argument/essay, just the key talking points about alternative options like having it become a no-op, or having it become a hard error.  A couple sentences is probably fine in this case.

Reviewer 2: This still needs to be addressed.

## Backward-Compatibility

This EIP requires a hard fork, since it modifies consensus rules.

> Few applications are affected by this change. The only use that breaks is where a contract is re-created at the same address using `CREATE2` (after a `SELFDESTRUCT`). An analysis of the contracts where this happens has been made available [here](https://hackmd.io/@albus/rkAbjAsWF). The only application that is significantly affected (and where code can be analyzed) is able to switch to a different model, and should have ample time to do so.

Reviewer 2: ```suggestion
Few applications are affected by this change. The only use that breaks is where a contract is re-created at the same address using `CREATE2` (after a `SELFDESTRUCT`).  The only application that is significantly affected (and where code can be analyzed) is able to switch to a different model, and should have ample time to do so.
```
No external links.  If the information is critical to the EIP it can be included inline or as an asset in `../assets/eip-4758/whatever.ext`.

---
eip: 4758
title: Deactivate SELFDESTRUCT
description: Deactivate SELFDESTRUCT by changing it to SENDALL, which does recover all funds to the caller but does not delete any code or storage.
author: Guillaume Ballet (@gballet), Vitalik Buterin (@vbuterin), Dankrad Feist (@dankrad)
discussions-to: https://ethereum-magicians.org/t/eip-4758-deactivate-selfdestruct/8710
status: Draft
type: Standards Track
category: Core
created: 2022-02-03
---

## Abstract

This EIP renames the `SELFDESTRUCT` opcode to `SENDALL`, and replaces its functionality. The new functionality will be only to send all Ether in the account to the caller.

## Motivation

The `SELFDESTRUCT` opcode requires large changes to the state of an account, in particular removing all code and storage. This will not be possible in the future with Verkle trees: Each account will be stored in many different account keys, which will not be obviously connected to the root account.

This EIP implements this change. Applications that only use `SELFDESTRUCT` to retrieve funds will still work.

## Specification

>  * The `SELFDESTRUCT` opcode is renamed to `SENDALL`, and now only immediately moves all ETH in the account to the target; it no longer destroys code or storage or alters the nonce

Reviewer 5: You forgot to mention that `SENDALL` still terminates execution.

>  * All refunds related to `SELFDESTRUCT` are removed

Reviewer 4: They were already removed in London.  Is there a reason to mention this here?

## Backward Compatibility

This EIP requires a hard fork, since it modifies consensus rules.

Few applications are affected by this change. The only use that breaks is where a contract is re-created at the same address using `CREATE2` (after a `SELFDESTRUCT`).

## Security Considerations

The following applications of `SELFDESTRUCT` will be broken and applications that use it in this way are not safe anymore:
1. Any use where `SELFDESTRUCT` is used to burn non-ETH token balances, such as ERC20, inside a contract. We do not know of any such use (since it can easily be done by sending to a burn address this seems an unlikely way to use `SELFDESTRUCT`)
2. Where `CREATE2` is used to redeploy a contract in the same place. There are two ways in which this can fail:
    - The destruction prevents the contract from being used outside of a certain context. For example, the contract allows anyone to withdraw funds, but `SELFDESTRUCT` is used at the end of an operation to prevent others from doing this. This type of operation can easily be modified to not depend on `SELFDESTRUCT`.

>     - The `SELFDESTRUCT` operation is used in order to make a contract upgradable. This is not supported anymore and delegates should be used.

Reviewer 6: This should say: this will break all the previously deployed code that depends on it.
